pkg/controllers/management/drivers/kontainerdriver/kontainerdriver.go

```diff
@@ -354,31 +328,58 @@ func (l *Lifecycle) Updated(obj *v3.KontainerDriver) (runtime.Object, error) {
 		return obj, l.updateDynamicSchema(obj)
 	}
 
+	var err error
+	var tmpObj runtime.Object
+	newObj := obj.DeepCopy()
+	switch {
+	case !obj.Spec.Active: // dealing with deactivate action
+		tmpObj, err = v3.KontainerDriverConditionInactive.DoUntilTrue(newObj, func() (runtime.Object, error) {
+			if err = l.dynamicSchemas.Delete(getDynamicTypeName(newObj), &v13.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
+				return nil, fmt.Errorf("error deleting schema: %v", err)
+			}
+			if err = l.removeFieldFromCluster(newObj); err != nil {
+				return nil, err
+			}
+
+			RemoveCondition(newObj, v3.KontainerDriverConditionActive)
+			return newObj, nil
+		})
+		newObj = tmpObj.(*v3.KontainerDriver)
+
 	// redownload file if active AND url changed or not downloaded
 	// prevents downloading on air-gapped installations
-	var err error
-	if obj.Spec.Active && (obj.Spec.URL != obj.Status.ActualURL || v3.KontainerDriverConditionDownloaded.IsFalse(obj) || !l.driverExists(obj)) {
-		obj, err = l.download(obj)
-		if err != nil {
-			return nil, err
+	case obj.Spec.URL != obj.Status.ActualURL ||
+		v3.KontainerDriverConditionDownloaded.IsFalse(obj) ||
+		!l.driverExists(obj):
+		if !v3.KontainerDriverConditionDownloaded.IsUnknown(newObj) {
+			v3.KontainerDriverConditionDownloaded.Unknown(newObj)
+			break
 		}
+		fallthrough
+	case !v3.KontainerDriverConditionDownloaded.IsTrue(newObj):
+		newObj, err = l.download(newObj)
+	case v3.KontainerDriverConditionActive.GetStatus(newObj) == "":
+		RemoveCondition(newObj, v3.KontainerDriverConditionInactive)
+		v3.KontainerDriverConditionActive.Unknown(newObj)
+	default:
+		tmpObj, err = v3.KontainerDriverConditionActive.DoUntilTrue(newObj, func() (runtime.Object, error) {
+			err = l.createDynamicSchema(newObj)
+			if err != nil {
+				return obj, err
+			}
+			return newObj, nil
+		})
+		newObj = tmpObj.(*v3.KontainerDriver)
 	}
-	// Create schema if obj is active
-	// Delete schema if obj is inactive
-	if obj.Spec.Active {
-		err = l.createDynamicSchema(obj)
 
-		v3.KontainerDriverConditionActive.True(obj)
-	} else {
-		if err = l.dynamicSchemas.Delete(getDynamicTypeName(obj), &v13.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
-			return nil, fmt.Errorf("error deleting schema: %v", err)
-		}
-		if err = l.removeFieldFromCluster(obj); err != nil {
-			return nil, err
-		}
+	if err != nil {
+		return nil, err
+	}
+	if !reflect.DeepEqual(obj, newObj) {
+		return l.kontainerDrivers.Update(newObj)
 	}
 
-	return obj, err
+	return obj, nil
```

Comment:
Maybe here we should return `newObj` instead of `obj`.

Reply:
No, we should return obj here because the obj is not modified.

Reply:
`newObj` will have its conditions modified inside `DoUntilTrue`.  See `condition.go` 208 and 231.